### PR TITLE
Delete alarms by id

### PR DESCRIPTION
**Pull request by molly-moen**
# Type of change? 

- [ ] Bug fix
- [x] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 

Add API to alarms to delete alarms by a list of alarm ids, up to a maximum of 1000 alarms. Also added single delete by id api.

Sample request:
POST .../telemetry/v1/alarms/delete
Body:
{
"Ids": ["977b5193-9e0f-410a-f57a-a1363cc4b87c", "a36d4da6-a0d9-366b-4fb9-f4d10439c2c5", "870f182d-3aac-bc07-6894-674bf7977192", "03d80e5f-93e7-205c-1074-7a51143d54d2"]
}

Response: 200 OK

This is the same content as https://github.com/Azure/device-telemetry-dotnet/pull/84
**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [x] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly
Will update wiki after check in.

### CODE REVIEW

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/runtime/AlarmsConfig.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+package com.microsoft.azure.iotsolutions.devicetelemetry.services.runtime;
+
```
Comment (isaac73) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/runtime/AlarmsConfig.java]: Add comments for class and methods

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Alarms.java [hunk 2]
```diff
@@ -168,6 +173,60 @@ public AlarmServiceModel update(String id, String status) throws Exception {
         return new AlarmServiceModel(document);
     }
 
+    @Override
+    public void delete(ArrayList<String> ids) throws Throwable {
```
Comment (isaac73) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Alarms.java]: Comments?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Alarms.java [hunk 3]
```diff
@@ -168,6 +173,60 @@ public AlarmServiceModel update(String id, String status) throws Exception {
...
+        CompletableFuture result = CompletableFuture.allOf(tasks);
+        try {
+            result.get();
+        } catch (ExecutionException ex) {
+            throw ex.getCause();
+        }
+    }
+
+    @Override
+    public void delete(String id) throws Exception {
```
Comment (isaac73) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Alarms.java]: Comments?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/AlarmsController.java [hunk 4]
```diff
@@ -76,4 +81,23 @@ public Result patch(String id) throws Exception {
 
         return ok(toJson(new AlarmApiModel(this.alarms.update(id, alarm.status))));
     }
+
+    public Result delete(String id) throws Exception {
```
Comment (isaac73) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/AlarmsController.java]: Comment referring to API URL will be helpful

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/AlarmsController.java [hunk 5]
```diff
@@ -76,4 +81,23 @@ public Result patch(String id) throws Exception {
...
+        if (id == null) {
+            return badRequest("no id given to delete");
+        }
+
+        this.alarms.delete(id);
+        return ok();
+    }
+
+    public Result deleteMultiple() throws Throwable {
```
Comment (isaac73) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/webservice/v1/controllers/AlarmsController.java]: Comment referring to API URL will be helpful

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Alarms.java [hunk 6]
```diff
@@ -168,6 +173,60 @@ public AlarmServiceModel update(String id, String status) throws Exception {
...
+        int retryCount = 0;
+        while (retryCount < this.maxDeleteRetryCount) {
+            try {
+                this.storageClient.deleteDocument(this.databaseName, this.collectionId, id);
+                return;
+
```
Comment (ppathan) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Alarms.java]: nit: extra line

File: test/services/test/AlarmsTest.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                "database",
+                "collection"),
+            new AlarmsConfig(
+                "documentdb",
+                "connString",
+                "database",
+                "collection",
+                3
+            ));
+        this.storageClient = Mockito.mock(IStorageClient.class);
```
Comment (ppathan) [inline test/services/test/AlarmsTest.java]: nit: storageClientMock?

File: app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Alarms.java [hunk 8]
```diff
@@ -168,6 +173,60 @@ public AlarmServiceModel update(String id, String status) throws Exception {
...
+                retryCount++;
+                if (retryCount >= this.maxDeleteRetryCount) {
+                    this.log.error("Failed to delete alarm " + id, e);
+                    throw new ExternalDependencyException(e.getMessage(), e);
+                }
+
+                this.log.warn("Exception on delete alarm " + id + ", retrying", e);
+                Thread.sleep(timeout);
+                throw new ExternalDependencyException(e.getMessage(), e);
```
Comment (ppathan) [inline app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Alarms.java]: When this line throws do we exit the while loop?

Comment (molly-moen) [reply app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Alarms.java]: you are right, I think I put that in for testing. Removing, it shouldn't throw there

**Pull request merged**